Introduces a guide on error handling strategies. (#328)
stellar/new-docs: content/docs/tutorials/handling-errors.mdx

#### Example Scenario
Suppose you submit a transaction and it enters the queue of the Stellar network, but Horizon crashes while giving you a response. Uncertain about the transaction status, you resubmit (with no changes!) until either (a) Horizon comes back up to give you a reply or (b) the timebounds you set get exceeded. There are only two possible results: the transaction makes it into a ledger (exactly once!) and Horizon gives you the response, or the transaction never makes it out of the queue and you receive the corresponding `tx_too_late` response.

#### Example Implementation

<CodeExample>

```js
let server = sdk.Server("horizon.stellar.org");

function submitTransaction(tx, timeout) {
  // If a maximum timebound isn't set, we should definitely set one.
  const maxTime = tx.maxTime || 20;
  tx.maxTime = maxTime;

  return server.submitTransaction(tx).catch(function (error) {
    if (isNonRetryErrorCase(error)) {
      // ...do other error handling...
      return;
    }
    
    timeout = timeout || 1; // start the back-off process
    return sleep(timeout).then(function () {
      return submitTransaction(tx, timeout + 5);
    });
  });
}
```

> (We assume the existence of a `sleep` implementation akin to the one [here](https://stackoverflow.com/a/39914235).)
> 
> </CodeExample>

Reviewer: Can't have text inside a `CodeExample`, I think this is probably what's causing the error
```suggestion
</CodeExample>

(We assume the existence of a `sleep` implementation akin to the one [here](https://stackoverflow.com/a/39914235).)
```

Author: Dunno how I overlooked this.. thank you!

#### Unsafe Transaction Adjustments
As outlined in the section on [retries](#retrying-until-success), resubmitting an *unchanged* (and valid) transaction (with the same operations, signatures, sequence number, etc.) is always safe to do. You should be careful when working around an error that *does* require changes to the transaction, though: it's very possible to cause **duplication transactions** which can result in all sorts of problems (double-payments, erroneous trustlines, etc.).

##### Example: Invalid Sequence Numbers
These errors typically occur when you have an outdated view of an account due to other transactions happening outside of your worldview. This could be because the account is used on multiple devices, you have concurrent submissions happening, or a number of other reasons. Thankfully, the solution is **usually** relatively simple: retrieve the account details and try again with an updated sequence number.

<CodeExample>

```js
// suppose `account` is an outdated `AccountResponse` object

> let tx = sdk.TransactionBuilder(account, ...)/* etc */.build()

Reviewer: The `...` here isn't valid JS, which might break some random syntax highlighting bits. Not the worst thing in the world, but thought I'd point it out while looking!

Author: Looks like it renders okay, but thanks for noting it! :+1:

Some errors cannot be overcome without changes to the request itself. 

#### Queries
Many of the `GET` requests have specific parameter requirements, and while the [SDKs](../software-and-sdks/) can help enforce them, you can still pass invalid arguments (e.g. an asset string that isn't [SEP-11 conformant](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0011.md#asset)) that error out every time. In this scenario, there's nothing you can do aside from following the [API specification](../../api) precisely. The `extras` field of the error response will often clue you in on where to look and what to look for:

<CodeExample>

```bash
curl -s https://horizon-testnet.stellar.org/claimable_balances/0000 | jq '.extras'
{
  "invalid_field": "id",
  "reason": "Invalid claimable balance ID"
}
```

</CodeExample>

Note that the SDKs make it a point to distinguish an **invalid request** (as above) vs. a **missing resource** (a `404 Not Found`) (e.g. the generic `NetworkError` vs. a `NotFoundError` in the JavaScript SDK), where the latter might not be considered an error depending on your situation.

#### Transaction Submissions

> Certain transaction submission failures also need adjustments to succeed. If the XDR is malformed or the transaction is somehow otherwise invalid, you'll encounter a `400 Bad Request` (for example, when excluding a source account as seen in the [API reference example](../../api/errors/response/)). Both transactions and their requisite operations can easily be malformed: ook at the `extras.result_codes` field for details and cross-reference them with the appropriate [Result Codes documentation](../../api/errors/result-codes/) to determine specifics.

Reviewer 2: ```suggestion
Certain transaction submission failures also need adjustments to succeed. If the XDR is malformed or the transaction is somehow otherwise invalid, you'll encounter a `400 Bad Request` (for example, when excluding a source account as seen in the [API reference example](../../api/errors/response/)). Both transactions and their requisite operations can easily be malformed: look at the `extras.result_codes` field for details and cross-reference them with the appropriate [Result Codes documentation](../../api/errors/result-codes/) to determine specifics.
```

Since there's no mechanism to "cancel" a transaction after it has been submitted, the first key to successful retries is leveraging [timebounds](https://developers.stellar.org/docs/glossary/transactions/#time-bounds): though optional, timebounds allow you to introduce a bit of determinism into an innately non-deterministic system. If the timebound has been exceeded, the transaction has a definitive, final state: either it made it into a ledger or it timed out. The second key comes from a certain Horizon guarantee: **you can safely retry submission as long as you don't modify the transaction**. Specifically,

> If the transaction has already been successfully applied to the ledger, Horizon will simply return the saved result and not attempt to submit the transaction again. Only in cases where a transaction’s status is unknown (and thus will have a chance of being included into a ledger) will a resubmission to the network occur.  

_(The above excerpt is from [this example](https://developers.stellar.org/docs/tutorials/send-and-receive-payments/#send-a-payment).)_

> In other words, if your transaction makes it through at any point in time, you'll get your previously-submitted transaction back for all subsequent retries.

Reviewer 2: ```suggestion
In other words, if your transaction makes it through at any point in time, you'll get your previously submitted transaction back for all subsequent retries.
```

#### Example Implementation

<CodeExample>

```js
let server = sdk.Server("horizon.stellar.org");

function submitTransaction(tx, timeout) {
  // If a maximum timebound isn't set, we should definitely set one.
  const maxTime = tx.maxTime || 20;
  tx.maxTime = maxTime;

  return server.submitTransaction(tx).catch(function (error) {
    if (isNonRetryErrorCase(error)) {
      // ...do other error handling...
      return;
    }
    
    timeout = timeout || 1; // start the back-off process
    return sleep(timeout).then(function () {
      return submitTransaction(tx, timeout + 5);
    });
  });
}
```

</CodeExample>

_(We assume the existence of a `sleep` implementation akin to the one [here](https://stackoverflow.com/a/39914235).)_

#### Details: Retry Backoff

> Be sure to integrate backoff into your retry mechanism. In our example error-handling code above, we implement a simple linear back-off, but there are [plenty of recommendations](https://backoff-utils.readthedocs.io/en/latest/strategies.html#why-are-backoff-strategies-useful) out there for various other strategies you can employ. Back-off is important both for maintaining performance and avoiding [rate-limiting](#rate-limiting) issues.

Reviewer 2: ```suggestion
Be sure to integrate backoff into your retry mechanism. In our example error-handling code above, we implement a simple linear backoff, but there are [plenty of recommendations](https://backoff-utils.readthedocs.io/en/latest/strategies.html#why-are-backoff-strategies-useful) out there for various other strategies you can employ. Backoff is important both for maintaining performance and avoiding [rate-limiting](#rate-limiting) issues.
```

When the Stellar network undergoes bursts of activity, [surge pricing](https://developers.stellar.org/docs/glossary/fees/#surge-pricing) might kick in. This and other fee fluctuations can cause unexpected errors during transaction submission in applications that don't plan ahead for its dynamic nature. A recent [blog post](https://www.stellar.org/developers-blog/transaction-submission-timeouts-and-dynamic-fees-faq) by the SDF discusses fee surges and other frequently asked questions in much greater detail. 

There are two main approaches for dealing with this variance:

  1. **Track fee fluctuations** via the [`fee_stats` endpoint](https://developers.stellar.org/api/aggregations/fee-stats/). This can let you make informed, specific choices about the fee you're comfortable paying. Alternatively, simply...

  2. **Set the highest fee** you are comfortable with. Crucially, you should remember that *this doesn't mean you'll **pay** that on every transaction*. You will only pay whatever is necessary to get you into the ledger: under normal (non-surge) circumstances, even with a higher *maximum* fee set, you will pay the *standard* fee (100 stroops as of this writing).

> The latter strategy balances simplicity, efficacy, and convenience, but unless you set your maximum high enough simply *never* be exceeded, can still lead to failures. The former strategy can provide more-reliable submissions by allowing tighter guarantees about whether or not a transaction will be accepted. In general, though, it's important to track fee costs: if the network saturates beyond your maximum willingness to pay, perhaps waiting for activity to die down or periodically retrying with the same fee is the best approach for your use case.

Reviewer 2: ```suggestion
The latter strategy balances simplicity, efficacy, and convenience, but unless you set your maximum high enough so that it's *never* exceeded, it can still lead to failures. The former strategy can provide more reliable submissions by allowing tighter guarantees about whether or not a transaction will be accepted. In general, though, it's important to track fee costs: if the network saturates beyond your maximum willingness to pay, perhaps waiting for activity to die down or periodically retrying with the same fee is the best approach for your use case.
```

order: 100
---

import { CodeExample } from "components/CodeExample";

When working with a payment network like Stellar, expecting the unexpected is critical in ensuring a good user experience: one in which money doesn't get lost, actions happen when they're supposed to (or as early as they can), and everyone's view of the network is as accurate as possible.

In many of the tutorials and code samples throughout this documentation, we've minimized the error handling code to reduce verbosity and focus on the essence of the examples. Here, we'll do the exact opposite, because error handling *is* the essence: by the end, you should be able to categorize errors and understand the best, most-idiomatic way to handle them in your application.

Since [the SDKs](../software-and-sdks/) are the main way people interact with the Stellar network, their error reporting will be the main focus of this document. At the end of the day, though, they're just human- and API-friendly abstractions on the underlying problems. As such, the knowledge should be relatively easy to transfer to other scenarios such as using a 3rd party SDK or interfacing with Horizon directly over HTTP.

> # Contents

Author: Good point; I didn't even realize there was an outline, which removes the need for the ToC.

<CodeExample>

```bash
curl -s --get https://horizon.stellar.org/transaction/e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855 | jq
```

</CodeExample>

<CodeExample>

```json
{
  "type": "https://stellar.org/horizon-errors/not_found",
  "title": "Resource Missing",
  "status": 404,
  "detail": "The resource at the url requested was not found.  This usually occurs for one of two reasons:  The url requested is not valid, or no data in our database could be found with the parameters provided."
}
```

</CodeExample>

It's worth noting that SDKs commonly classify 404s uniquely relative to other error codes (e.g. `NotFoundError` vs the generic `NetworkError` in the JavaScript SDK).

## Resolutions
TODO: Either fix the parameters (in case of user error, insert link to API reference) or try again (in case of transient error). 

Are there other cases?

> # Transaction Submission Errors

Reviewer 3: I think it's worth adding a general rule of thumb here: any response other than `200 OK` or `400 Bad Request` means that the final status of the transaction is unknown.

Also, would be great to create a section with a generic error handling algorithm in pseudocode.

Author: > Also, would be great to create a section with a generic error handling algorithm in pseudocode.

TBD in a future PR!

  "title": "Transaction Failed",
  "status": 400,
  "detail": "The transaction failed when submitted to the stellar network. The `extras.result_codes` field on this response contains further details.  Descriptions of each code can be found at: https://www.stellar.org/developers/guides/concepts/list-of-operations.html",
  "extras": {
    "envelope_xdr": "AAAAAgAAAABjsEXePrFn+dH3tNaTkGo3kojf2iecJqU3g84/95XLqgAAAGQAAAAAbG1hbwAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAA50aGlzIGlzIGEgdGVzdAAAAAAAAQAAAAEAAAAAY7BF3j6xZ/nR97TWk5BqN5KI39onnCalN4POP/eVy6oAAAAKAAAAB21lc3NhZ2UAAAAAAQAAAA1IZWxsbywgd29ybGQhAAAAAAAAAAAAAAH3lcuqAAAAQNZ/9lkhKMw0Rbiwmw+dF3X1DldjT0qPdiCuEPpf5rErYDoXBkr0HlQdAfK1ptfLwK7kMELQfXCfPUEYVuJd4gQ=",
    "result_codes": {
      "transaction": "tx_bad_seq"
    },
    "result_xdr": "AAAAAAAAAAD////7AAAAAA=="
  }
}
```

</CodeExample>

## Invalid Sequences
These errors typically occur when you have an outdated view of an account due to other transactions happening outside of your worldview. This could be because the account is used on multiple devices, you have concurrent submissions happening, or a number of other reasons.

> Thankfully, the solution is relatively simple: retrieve the account details and try again with an updated sequence number.

Reviewer 3: I think it's worth noting here that updating sequence number to the latest value can sometime cause a double spend. Say you have two submitters running and they both try to send the same payment tx with the same seqnum. One of them will succeed, the other will get bad sequence error and will try to submit exactly the same tx again. If it succeeds this time the payment will be sent twice.

// when submitting any transaction, first query fee stats
server.feeStats().then(function (response) {
  let tx = base.TransactionBuilder(someAccount, {
    fee: parseFloat(response.fee_charged.p50) * 1.10).toFixed(2),
    // ...
  });
  // ...build the rest of the tx... 
  tx.sign(someAccount);
  return server.submitTransaction(tx);
});
```

</CodeExample>

### Bumping fees on past transactions
It's possible that even with the liberal fee-paying policy, your transaction fails to make it into the ledger due to insufficient funds or untimely surges. Resolving this problem is exactly the goal of a [fee-bump transaction](https://developers.stellar.org/docs/glossary/fee-bumps/); the following snippet shows how you can resubmit a transaction with a higher fee given that you have the original [transaction envelope](https://developers.stellar.org/docs/glossary/transactions/#transaction-envelopes):

<CodeExample>

```js
// Let `lastTx` be some transaction that fails submission due to high fees, and
// `lastFee` be the maximum fee (expressed as a float) willing to be paid by 
// `account` for `lastTx`.
server.submitTransaction(lastTx).catch(function (error) {
  if (isFeeError(error)) {

>     let bump = sdk.TransactionBuilder.buildFeeBumpTransaction(
>       account, (lastFee * 10).toFixed(2), lastTx, server.networkPassphrase
>     );

Reviewer 3: This isn't clear to me. The fee should be an integer (in stroops) but we use `toFixed` here. Also adding a suggestion that will be more readable:
```suggestion
    let bump = sdk.TransactionBuilder.buildFeeBumpTransaction(
      account, // Account that will pay the fee
      (lastFee * 10).toFixed(2), // New fee
       lastTx,
       server.networkPassphrase
    );
```

Author: Good point! It's supposed to be an integer but encoded as a string; I'll rectify this! `.toFixed` is necessary to avoid long decimals (e.g. `1000.000000001`) but you're right that there shouldn't be a decimal in the first place.

There's no mechanism to "cancel" a transaction after it has been submitted aside from by exceeding the timebounds. In fact, one of the best tools available for managing transaction failures is through [timebounds](https://developers.stellar.org/docs/glossary/transactions/#time-bounds). While optional, they're *crucial* in simplifying error handling: it's much easier to know when to check a transaction's status when you know the maximum amount of time it can be stuck.

With this trick in our back pocket, resolving 504s is straightforward: until the timebound has expired, you can safely retry submission *without modifying the transaction*. Horizon makes an important guarantee (excerpt is from [this example](https://developers.stellar.org/docs/tutorials/send-and-receive-payments/#send-a-payment)) for this scenario:

> If the transaction has already been successfully applied to the ledger, Horizon will simply return the saved result and not attempt to submit the transaction again. Only in cases where a transaction’s status is unknown (and thus will have a chance of being included into a ledger) will a resubmission to the network occur.

> In other words, if your transaction made it through after the 504, you'll get your previously-submitted transaction back. If it didn't, you've essentially started from scratch again: you might 504 again, fail for another reason (e.g. `tx_bad_seq`), or succeed.

Reviewer 3: ```suggestion
In other words, if your transaction made it through after the 504, you'll get your previously-submitted transaction's result back. If it didn't, you've essentially started from scratch again: you might 504 again, fail for another reason (e.g. `tx_bad_seq`), or succeed.
```

Author: Good point!

## Outages
Receiving no response from Horizon after a submission is very similar to receiving a 504: you're uncertain about the outcome of the transaction submission.

> It's obviously crucial to determine whether or not it went through or not, but you can likewise just keep trying submission until the timebounds expire. Once they have, though, if the transient error hasn't resolved itself, you should just stop and wait; submitting a new transaction might result in a undesirable duplicate action. 

Reviewer 3: > just keep trying submission until the timebounds expire

I'm not sure if we want to complicate it more but _keep trying_ no longer works with banning in stellar-core. So after a few ledgers the only option may be to send fee bump tx.

Author: You need to update timebounds even w/ a fee bump, right? So the txhash would differ. So you may cause a double-spend? Just making sure this is accurate so I note it.

Reviewer 3: I'm talking about the case in which `timebounds.max_time` is still in the future. When it's in the past and tx is not in the ledger you can create a new tx with the same seqnum.

Here are some of the common errors encountered when performing a `POST /transactions`:

  - [400 errors, e.g. `tx_bad_seq`](#invalid-sequences)
  - [Fee / surge pricing errors, e.g. `tx_insufficient_fee`](#surge-pricing)
  - [504 error, timeout](#timeouts)
  - [No HTTP response](#outages)

Keep in mind this general rule of thumb: **any response other than `200 OK` or `400 Bad Request` means that the final status of the transaction is unknown.**

### 400 Errors
Generally-speaking, if you get a 400 error during transaction submission, there are two main causes: either your transaction is in an unacceptable format or Stellar Core rejected your transaction. The former case is described by the [`transaction_malformed`](https://developers.stellar.org/api/errors/http-status-codes/horizon-specific/transaction-malformed/) error code and the latter by the [`transaction_failed`](https://developers.stellar.org/api/errors/http-status-codes/horizon-specific/transaction-failed/) code.

> Some transaction failures are recoverable; you should look at the `extras.result_code` field in the response message to determine the [root cause](https://developers.stellar.org/api/errors/result-codes/transactions/). One of these recoverable errors is `tx_bad_seq`, where the sequence number on the transaction doesn't match what's expected for the account; another is `tx_insufficient_fee` which is covered [later](#surge-pricing).

Author: Good point, but I want to avoid enumerating errors if possible, since that should be *more-or-less* researchable from the [`result_codes` page itself](https://developers.stellar.org/api/errors/result-codes/transactions/). I don't think it's possible to be exhaustive. You're right, though; I'll think about this some more!

> If you submit two distinct transactions with the same source account and sequence number, and the second transaction is a fee-bump transaction, the second transaction will be included in the transaction queue in place of the first transaction if and only if **the fee bid of the second transaction is at least 10x the fee bid of the first transaction**.

### Timeouts
If you receive a 504 error from Horizon after a transaction submission, you've encountered a timeout. In the Stellar universe, a timeout indicates something a little nebulous and non-traditional: it's more of a *warning* that your request hasn't been fulfilled within a reasonable amount of time *yet* more than an error. This subtlety arises because of the nature of the relationship between Horizon and Stellar Core: the network might take some time to accept a transaction---especially one with a low fee, see [Surge Pricing](#surge-pricing) later---on the order of 5-10 minutes in the worst cases, whereas Horizon needs to provide developers with a response within 30 seconds or so.

> This leads to a very important point: **receiving a 504 for your transaction submission does not mean the transaction did not make it to the network**.

Reviewer 4: TIL x2! Great info here 💯 

question @Shaptic , if you don’t set timebounds on a transaction what’s the upperbound on the time Core will try to add your transaction to the ledger? (eg. if theoretically there are always competing transactions with higher baseFees during surge pricing, when does Core give up on your transaction)

I see you put `5-10 minutes` is there a hard max value for that? Either way I think adding a link or short note on Core's max effort for your transaction would be a helpful note here

> This leads to a very important point: **receiving a 504 for your transaction submission does not mean the transaction did not make it to the network**.

Author: I think you got your answer in Slack @acharb, but for posterity, it appears that transactions w/o timebounds will **typically** be considered for 3 ledgers then banned for 10, but this can go on forever.

There's no mechanism to "cancel" a transaction after it has been submitted aside from by exceeding its timebounds. In fact, one of the best tools available for managing transaction failures is through [timebounds](https://developers.stellar.org/docs/glossary/transactions/#time-bounds). While optional, they're *crucial* in simplifying error handling: it's much easier to know when to check a transaction's status when you know the maximum amount of time it can be stuck.

With this trick in our back pocket, resolving 504s is straightforward: until the timebound has expired, you can safely retry submission *without modifying the transaction*. Horizon makes an important guarantee (excerpt is from [this example](https://developers.stellar.org/docs/tutorials/send-and-receive-payments/#send-a-payment)) for this scenario:

> If the transaction has already been successfully applied to the ledger, Horizon will simply return the saved result and not attempt to submit the transaction again. Only in cases where a transaction’s status is unknown (and thus will have a chance of being included into a ledger) will a resubmission to the network occur.

> In other words, if your transaction made it through after the 504, you'll get your previously-submitted transaction back. If it didn't (e.g. a `tx_too_late` result code), you're essentially back to square one: you might 504 again, fail for another reason (e.g. `tx_bad_seq`), or succeed.

Reviewer 5: I didn't realize `504` meant core was still deciding -- I figured it had been rejected and that was that.

This brings me to another question. Lets say I get a 504. This means my fee is above the minimum so I didn't get a `tx_insufficient_fee`, but its not competitive with the other tx's being submitted, right? Is there a non-fee-related issue that would result in getting a `504` back?

If not, then we should state that `504`s mean the transaction's was too low relative to the other transactions being submitted, and maybe explain how that differs from `tx_insufficient_fee`. 

If there are other scenarios, is there a way to figure out what the reason was, if it wasn't related to fees?

> In other words, if your transaction made it through after the 504, you'll get your previously-submitted transaction back. If it didn't (e.g. a `tx_too_late` result code), you're essentially back to square one: you might 504 again, fail for another reason (e.g. `tx_bad_seq`), or succeed.

Author: > Is there a non-fee-related issue that would result in getting a 504 back?

Sure, there are plenty of transient issues that could cause it, I think. From my understanding, a `504 Timeout` **might** mean that Core is still deciding; it could also be due to a problem between Horizon and Core for example, or the Core node and the rest of the Stellar network. 

It's hard (if not impossible) to determine those error details, which is why tx resub until a hard success/failure is basically the only thing you can do.

#### Transaction Bans
Note the following caveat of the [transaction lifecycle](https://developers.stellar.org/docs/glossary/transactions/#transaction-lifecycle): 

> > If a transaction is kept in memory after a certain number of ledger closes, **it will be banned for several additional ledgers.** This means no attempt will be made to include it in a candidate transaction set additional ledgers during this time.

Reviewer 5: How does one end up causing this situation? If I get a `504`, and I keep resubmitting, am I risking the transaction getting banned?

Author: From my understanding, bans happen if a tx stays in the queue for too long (others are taking priority due to higher fees). I've clarified this!

Reviewer 5: So as I understand, a transaction will stay in the queue until one of the following happens:
- it is included in the ledger (success)
- its upper timebound is exceeded (tx_too_late)
- its banned

Please correct me if I'm wrong but this raises the following questions:
- is there a fixed limit on how long a tx can stay in the queue (# of ledgers / # of seconds)? This limit should be mentioned if it exists.
- what is the Horizon response when I try to submit a banned transaction? I don't see a result code that directly applies
- Lets say every transaction submitted has a higher fee than the one I submit, and I don't include an upper timebound. The TX _will_ be banned right?

Reviewer 5: Oh I see you answered my first question here: https://github.com/stellar/new-docs/pull/328#discussion_r579516471

But you say "typically". If I want to avoid getting banned (because I don't want to pay the 10x required to be reconsidered), what guidance would you give me? Always set an upper timebound of no more than 15 seconds (~3 ledgers)? I think we should provide some guidance here in the doc, because getting banned is a significant penalty.

Author: I believe there's no limit on staying in the queue aside from what's achievable via the tx timebound. Here's what @MonsieurNicolas had to say on the matter:

> You need to set your timeouts based on your business requirements and assume that transactions may get dropped or re-processed at any time (transactions may sit around in queues for much longer than 3 ledgers). The only thing you can rely on are close time in the latest ledger that was closed and sequence number of your account in that latest ledger, then you need to design a sequence of transactions that deals with those races properly.

If you want to avoid a ban, the best strategy is (_likely_-I'm doing some inferring here) just to set your fees appropriately. For example, I can't imagine setting a max fee based on the 90th percentile (from `/fee_stats`) will ever lead to getting banned. I'll try to get answers for your other bullets from Core.

Author: Despite our mutual curiosity and desire for clarity, @JakeUrban, I think it's actually best to avoid any mention of tx bans in this guide entirely (see fd87d6b68cb7f4ccd6748c58f57c05dd6d94e20d).

This thought process is magnified by the fact that a txban-specific error doesn't bubble up to SDK users and so represents an error scenario that's fairly deep and specific. As far as users are concerned, the resolution strategy stays the same: **set the max fee you're willing to pay** (or make a fee-bump / check `/fee_stats`) **and try again**. I think "exposing" this (deep) knowledge about Stellar Core's behavior (a) isn't actually helpful (except to satisfy curiosity) (b) strips away more of the (intended) abstraction of Horizon <--> Core and (c) locks Core into certain expectations. 

What do you think?

Reviewer 5: I don't think we need to delay merging this doc over this conversation, but I do think it exposed some gaps in our understanding. 

It's not clear how Horizon handles Core's `TRY_AGAIN_LATER` error and how that is relayed to the client. It may be the case that Horizon lumps in into other error modes described in this doc (`504`), in which case the client truly doesn't need to worry about it. But I wanted to confirm that there isn't some undocumented error response that represents a tx ban that clients wouldn't be prepared to handle.
